library/core/src/array/mod.rs

```diff
@@ -404,6 +405,91 @@ array_impl_default! {32, T T T T T T T T T T T T T T T T T T T T T T T T T T T T
 
 #[lang = "array"]
 impl<T, const N: usize> [T; N] {
+    /// Creates an array `[T; N]` where each array element `T` is returned by the `cb` call.
+    ///
+    /// # Arguments
+    ///
+    /// * `cb`: Callback where the passed argument is the current array index.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// #![feature(generate_array)]
+    ///
+    // let array = <[usize; 5]>::generate(|i| i);
+    // assert_eq!(array, [0, 1, 2, 3, 4]);
+    /// ```
+    #[inline]
+    #[unstable(feature = "generate_array", issue = "none")]
+    pub fn generate<F>(mut cb: F) -> [T; N]
+    where
+        F: FnMut(usize) -> T,
+    {
+        let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+        let mut guard: Guard<T, N> =
+            Guard { dst: MaybeUninit::slice_as_mut_ptr(&mut array), initialized: 0 };
+        for (idx, value) in array.iter_mut().enumerate() {
+            value.write(cb(idx));
+            guard.initialized += 1;
+        }
+        // FIXME: Convert to crate::mem::transmute once it works with generics.
+        mem::forget(guard);
+        // SAFETY: It is guaranteed that the whole array will be filled
+        unsafe { mem::transmute_copy::<_, [T; N]>(&array) }
```

Comment:
Now that we have `[T; N]::map()` from #75212, what about something along the lines of this ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=76b2da3da325b7c8839b96695d002508))?
```suggestion
        let mut idx = 0;
        [(); N].map(|_| {
            let res = cb(idx);
            idx += 1;
            res
        })
```

Reply:
Both use the same underlying function and generate the same assembly output (https://godbolt.org/z/jqj5rrzdv) so OK! One less `unsafe`.